rle_array/_slicing.py

```diff
@@ -0,0 +1,395 @@
+"""
+Helpers that allows us to deal with Python slicing.
+
+The issue with Python slicing are:
+
+- ``slice`` type:
+  - the types in ``slice`` are completely unchecked (can even be a string or any user-provided type)
+  - the consistency of the values in ``slice`` are unchecked
+  - there is not information about the container size (which makes consistency checks more complicated)
+
+- ``slice.step`` value:
+  - ``step`` has the implicit default 1
+  - there can be forward and backward slices depending on the ``step`` value
+  - there can be step sizes which are not modulo 1
+
+- ``slice.start`` and ``slice.stop`` values:
+  - the implicit defaults of ``start`` and ``stop`` depend on ``step`` (is it positive or negative?)
+  - ``start`` and ``stop`` can be negative (aka "from the end")
+  - ``start`` and ``stop`` can over/underflow the container
+
+We do not want to deal with all these edge cases in every code snipped that deals with slicing, so we introduce
+:class:`NormalizedSlice` that solves the issue in a central place.
+"""
+from typing import Optional
+
+import numpy as np
+
+
+class NormalizedSlice:
+    """
+    A normalized slice.
+
+    .. important::
```

Comment:
Nitpick: You tell the users to not construct this class by hand. But in your tests you construct the test cases by hand which causes you to duplicate a lot of sanity checks(in your class and in the from slice method)) It looks like otherwise parts of the validation code is actually unreachable.

Reply:
This doesn't mean that `from_slice` and `project` do not have any bugs. The double check here is helpful for that. Also, it makes it simpler to construct correct slices for test purposes. Keep in mind the the downstream consumers for `NormalizedSlice` blindly rely on the correct implementation. So I would rather opt for double checking than for hard-to-debug issues.